File: frontend/lib/intermediaries-filing/overview.ts
/**
 * Overview summary derived from the in-memory Intermediaries & Filing draft (IF1).
 */

import {
  assessIntermediariesFiling,
  type IfAssessmentResponse,
} from '@/lib/intermediaries-filing/assessment';
import {
  computeIntermediariesFilingModel,
  type IntermediariesFilingModel,
} from '@/lib/intermediaries-filing/compute';
import { IF_SECTION_LABELS } from '@/lib/intermediaries-filing/options';
import { calculateIntermediariesFilingProgress } from '@/lib/intermediaries-filing/progress';
import {
  createEmptyLinkedWorkstreamReferences,
  type IfProgress,
  type LinkedWorkstreamReferences,
} from '@/lib/intermediaries-filing/types';
import type {
  IntermediariesFilingPayload,
  IntermediariesFilingSectionId,
} from '@/lib/schemas/intermediaries-filing';

export type IntermediariesFilingOverviewSummary = {
  sectionStatuses: IfProgress['sections'];
  sectionsComplete: number;
  sectionsInProgress: number;
  totalSections: number;
  overallStatus: IfProgress['overallStatus'];
  currentFilingStage: string;
  targetSmePlatform: string;
  issueMethod: string;
  freshIssueAmount: string;
  ofsAmount: string;
  totalOfferAmount: string;
  currentPriceBandStatus: string;
  authoritativeDocumentVersion: string;
  intermediaryCount: number;
  leadManagerCount: number;
  activeIntermediaryCount: number;
  agreementsPendingCount: number;
  registrationsPendingReview: number;
  ddAreasSignedOff: number;
  ddAreasTotal: number;
  openDdAreas: number;
  certificatesReady: number;
  certificatesPending: number;
  consentsRequired: number;
  consentsReceived: number;
  chapterSignoffsComplete: number;
  chapterSignoffsTotal: number;
  filingCount: number;
  openExchangeQueries: number;
  overdueExchangeQueries: number;
  inPrincipleStatus: string;
  sebiSmeFilingStatus: string;
  rocFilingStatus: string;
  isinStatus: string;
  sponsorBankReady: boolean;
  upiReady: boolean;
  asbaReady: boolean;
  bankRolesReady: number;
  bankRolesTotal: number;
  underwritingCoverage: string;
  uncoveredShares: string;
  merchantBankerOwnAccountPercentage: string;
  marketMakerAppointed: boolean;
  marketMakingAgreementExecuted: boolean;
  marketMakingReservationStatus: string;
  issueOpeningDate: string;
  issueClosingDate: string;
  preliminaryTPlus3ListingDate: string;
  basisStatus: string;
  dematStatus: string;
  listingStatus: string;
  unresolvedPlaceholders: number;
  inspectionItemsPending: number;
  issueAgreementsPending: number;
  advertisementsPending: number;
  repositoryReadiness: string;
  reconciliationMismatchCount: number;
  assessmentConcerns: number;
  pendingProfessionalConfirmations: number;
  assessmentResult: IfAssessmentResponse['result'];
  assessmentResultLabel: string;
  assessmentSummary: string;
  recommendedNextActions: Array<{ sectionId: IntermediariesFilingSectionId; label: string }>;
};

export function buildOverviewSummary(
  payload: IntermediariesFilingPayload,
  linkedReferences: LinkedWorkstreamReferences = createEmptyLinkedWorkstreamReferences(),
): IntermediariesFilingOverviewSummary {
  const progress = calculateIntermediariesFilingProgress(payload);
  const model = computeIntermediariesFilingModel(payload, linkedReferences);
  const assessment = assessIntermediariesFiling(payload, linkedReferences);
  const config = payload.issueConfigurationAndFilingSnapshot;
  const section3 = payload.filingAndRegulatoryMilestoneTracker;
  const section8 =
    payload.finalOfferDocumentAdvertisementsMaterialDocumentsAndFilingReadiness;

  const sectionsInProgress = Object.values(progress.sections).filter(
    (status) => status === 'in_progress',
  ).length;

  const incompleteSections = (
    Object.entries(progress.sections) as Array<
      [IntermediariesFilingSectionId, IfProgress['sections'][IntermediariesFilingSectionId]]
    >
  ).filter(([, status]) => status !== 'complete' && status !== 'not_applicable');

  const recommendedNextActions = incompleteSections.slice(0, 4).map(([sectionId]) => ({
    sectionId,
    label: `Continue with ${IF_SECTION_LABELS[sectionId]}`,
  }));

  const assessmentConcerns =
    assessment.counts.potentialConcern +
    assessment.counts.exchangeQueryPending +
    assessment.counts.underwritingPending +
    assessment.counts.marketMakingPending +
    assessment.counts.issueInfrastructurePending +
    assessment.counts.listingActionPending;

  return {
    sectionStatuses: progress.sections,
    sectionsComplete: progress.sectionsComplete,
    sectionsInProgress,
    totalSections: progress.totalSections,
    overallStatus: progress.overallStatus,
    currentFilingStage: progress.currentFilingStage,
    targetSmePlatform:
      config.ipoSetupLinkedSnapshot.targetSmePlatform ||
      linkedReferences.ipoSetup.targetSmePlatform ||
      '',
    issueMethod:
      config.ipoSetupLinkedSnapshot.issueMethod || linkedReferences.ipoSetup.issueMethod || '',
    freshIssueAmount: config.filingSnapshotReconciliation.freshIssueAmount,
    ofsAmount: config.filingSnapshotReconciliation.ofsAmount,
    totalOfferAmount: config.filingSnapshotReconciliation.totalOfferAmount,
    currentPriceBandStatus: config.pricing.priceBand || config.pricing.finalIssuePrice,
    authoritativeDocumentVersion: model.finalDocumentAggregates.authoritativeVersionLabel,
    intermediaryCount: model.intermediaryAggregates.totalCount,
    leadManagerCount: model.intermediaryAggregates.leadManagerCount,
    activeIntermediaryCount: model.intermediaryAggregates.activeCount,
    agreementsPendingCount: model.intermediaryAggregates.agreementPendingCount,
    registrationsPendingReview: model.intermediaryAggregates.registrationsPendingReview,
    ddAreasSignedOff: model.dueDiligenceAggregates.signedOffCount,
    ddAreasTotal: model.dueDiligenceAggregates.areaCount,
    openDdAreas:
      model.dueDiligenceAggregates.areaCount - model.dueDiligenceAggregates.signedOffCount,
    certificatesReady: model.certificateConsentAggregates.signedCertificateCount,
    certificatesPending: model.certificateConsentAggregates.certificatesPending,
    consentsRequired: model.certificateConsentAggregates.consentCount,
    consentsReceived: model.certificateConsentAggregates.consentsReceived,
    chapterSignoffsComplete: model.certificateConsentAggregates.chapterSignoffsComplete,
    chapterSignoffsTotal: model.certificateConsentAggregates.chapterSignoffsTotal,
    filingCount: model.filingAggregates.filingCount,
    openExchangeQueries: model.filingAggregates.openQueryCount,
    overdueExchangeQueries: model.filingAggregates.overdueQueryCount,
    inPrincipleStatus: section3.inPrincipleApproval.approvalReceived,
    sebiSmeFilingStatus: section3.sebiSmeFiling.status,
    rocFilingStatus: section3.rocFiling.filingComplete,
    isinStatus: model.infrastructureAggregates.isinStatus,
    sponsorBankReady: model.infrastructureAggregates.sponsorBankReady,
    upiReady: model.infrastructureAggregates.upiReady,
    asbaReady: model.infrastructureAggregates.asbaReady,
    bankRolesReady: model.infrastructureAggregates.configuredBankRoles,
    bankRolesTotal: model.infrastructureAggregates.requiredBankRoles,
    underwritingCoverage: model.underwritingAggregates.totalUnderwritingPercentage,
    uncoveredShares: model.underwritingAggregates.uncoveredShares,
    merchantBankerOwnAccountPercentage: model.underwritingAggregates.ownAccountPercentage,
    marketMakerAppointed: model.marketMakingAggregates.marketMakerAppointed,
    marketMakingAgreementExecuted: model.marketMakingAggregates.agreementExecuted,
    marketMakingReservationStatus:
      payload.underwritingMarketMakingAndDistributionArrangements.marketMakerReservation
        .allocationStatus,
    issueOpeningDate: model.programmeAggregates.issueOpeningDate,
    issueClosingDate: model.programmeAggregates.issueClosingDate,
    preliminaryTPlus3ListingDate: model.programmeAggregates.preliminaryTPlus3ListingDate,
    basisStatus:
      payload.issueProgrammeAllotmentListingAndPostIssueExecution.basisOfAllotment.basisPrepared,
    dematStatus:
      payload.issueProgrammeAllotmentListingAndPostIssueExecution.dematCredit.sharesCredited,
    listingStatus:
      payload.issueProgrammeAllotmentListingAndPostIssueExecution.listing.listingCompletionStatus,
    unresolvedPlaceholders: model.finalDocumentAggregates.openPlaceholderCount,
    inspectionItemsPending: model.finalDocumentAggregates.inspectionItemsPending,
    issueAgreementsPending: model.finalDocumentAggregates.issueAgreementsPending,
    advertisementsPending: model.finalDocumentAggregates.publicCommunicationsPending,
    repositoryReadiness:
      section8.merchantBankerDdRepositoryReadiness.uploadComplete === 'yes'
        ? 'complete'
        : section8.merchantBankerDdRepositoryReadiness.uploadProcessStarted === 'yes'
          ? 'in_progress'
          : 'not_started',
    reconciliationMismatchCount: model.reconciliation.totalMismatchCount,
    assessmentConcerns,
    pendingProfessionalConfirmations: assessment.counts.pendingProfessionalConfirmation,
    assessmentResult: assessment.result,
    assessmentResultLabel: assessment.resultLabel,
    assessmentSummary: assessment.summary,
    recommendedNextActions,
  };
}

export type { IntermediariesFilingModel };

File: backend/app/modules/intermediaries_filing/overview.py
"""Overview summary derived from Intermediaries & Filing draft (IF2)."""

from __future__ import annotations

from typing import Any

from app.modules.intermediaries_filing.assessment import assess_intermediaries_filing
from app.modules.intermediaries_filing.compute import compute_intermediaries_filing_model
from app.modules.intermediaries_filing.constants import SECTION_LABELS
from app.modules.intermediaries_filing.progress import calculate_intermediaries_filing_progress


def build_overview_summary(
    payload: dict[str, Any],
    linked_references: dict[str, Any] | None = None,
) -> dict[str, Any]:
    linked = linked_references or {}
    progress = calculate_intermediaries_filing_progress(payload)
    model = compute_intermediaries_filing_model(payload, linked)
    assessment = assess_intermediaries_filing(payload, linked)
    config = payload.get("issueConfigurationAndFilingSnapshot") or {}
    section3 = payload.get("filingAndRegulatoryMilestoneTracker") or {}
    section8 = payload.get("finalOfferDocumentAdvertisementsMaterialDocumentsAndFilingReadiness") or {}
    ipo_snapshot = config.get("ipoSetupLinkedSnapshot") or {}
    ipo_linked = linked.get("ipoSetup") or {}

    sections_in_progress = sum(1 for status in progress["sections"].values() if status == "in_progress")
    incomplete_sections = [
        section_id
        for section_id, status in progress["sections"].items()
        if status not in {"complete", "not_applicable"}
    ]
    recommended_next_actions = [
        {
            "sectionId": section_id,
            "label": f"Continue with {SECTION_LABELS.get(section_id, section_id)}",
        }
        for section_id in incomplete_sections[:4]
    ]

    assessment_concerns = (
        assessment["counts"]["potentialConcern"]
        + assessment["counts"]["exchangeQueryPending"]
        + assessment["counts"]["underwritingPending"]
        + assessment["counts"]["marketMakingPending"]
        + assessment["counts"]["issueInfrastructurePending"]
        + assessment["counts"]["listingActionPending"]
    )

    repo = section8.get("merchantBankerDdRepositoryReadiness") or {}
    if repo.get("uploadComplete") == "yes":
        repository_readiness = "complete"
    elif repo.get("uploadProcessStarted") == "yes":
        repository_readiness = "in_progress"
    else:
        repository_readiness = "not_started"

    reconciliation = config.get("filingSnapshotReconciliation") or {}
    section6 = payload.get("underwritingMarketMakingAndDistributionArrangements") or {}
    section7 = payload.get("issueProgrammeAllotmentListingAndPostIssueExecution") or {}

    return {
        "sectionStatuses": progress["sections"],
        "sectionsComplete": progress["sectionsComplete"],
        "sectionsInProgress": sections_in_progress,
        "totalSections": progress["totalSections"],
        "overallStatus": progress["overallStatus"],
        "currentFilingStage": progress["currentFilingStage"],
        "targetSmePlatform": ipo_snapshot.get("targetSmePlatform") or ipo_linked.get("targetSmePlatform") or "",
        "issueMethod": ipo_snapshot.get("issueMethod") or ipo_linked.get("issueMethod") or "",
        "freshIssueAmount": reconciliation.get("freshIssueAmount", ""),
        "ofsAmount": reconciliation.get("ofsAmount", ""),
        "totalOfferAmount": reconciliation.get("totalOfferAmount", ""),
        "currentPriceBandStatus": (config.get("pricing") or {}).get("priceBand")
        or (config.get("pricing") or {}).get("finalIssuePrice")
        or "",
        "authoritativeDocumentVersion": model["finalDocumentAggregates"]["authoritativeVersionLabel"],
        "intermediaryCount": model["intermediaryAggregates"]["totalCount"],
        "leadManagerCount": model["intermediaryAggregates"]["leadManagerCount"],
        "activeIntermediaryCount": model["intermediaryAggregates"]["activeCount"],
        "agreementsPendingCount": model["intermediaryAggregates"]["agreementPendingCount"],
        "registrationsPendingReview": model["intermediaryAggregates"]["registrationsPendingReview"],
        "ddAreasSignedOff": model["dueDiligenceAggregates"]["signedOffCount"],
        "ddAreasTotal": model["dueDiligenceAggregates"]["areaCount"],
        "openDdAreas": model["dueDiligenceAggregates"]["areaCount"]
        - model["dueDiligenceAggregates"]["signedOffCount"],
        "certificatesReady": model["certificateConsentAggregates"]["signedCertificateCount"],
        "certificatesPending": model["certificateConsentAggregates"]["certificatesPending"],
        "consentsRequired": model["certificateConsentAggregates"]["consentCount"],
        "consentsReceived": model["certificateConsentAggregates"]["consentsReceived"],
        "chapterSignoffsComplete": model["certificateConsentAggregates"]["chapterSignoffsComplete"],
        "chapterSignoffsTotal": model["certificateConsentAggregates"]["chapterSignoffsTotal"],
        "filingCount": model["filingAggregates"]["filingCount"],
        "openExchangeQueries": model["filingAggregates"]["openQueryCount"],
        "overdueExchangeQueries": model["filingAggregates"]["overdueQueryCount"],
        "inPrincipleStatus": (section3.get("inPrincipleApproval") or {}).get("approvalReceived", ""),
        "sebiSmeFilingStatus": (section3.get("sebiSmeFiling") or {}).get("status", ""),
        "rocFilingStatus": (section3.get("rocFiling") or {}).get("filingComplete", ""),
        "isinStatus": model["infrastructureAggregates"]["isinStatus"],
        "sponsorBankReady": model["infrastructureAggregates"]["sponsorBankReady"],
        "upiReady": model["infrastructureAggregates"]["upiReady"],
        "asbaReady": model["infrastructureAggregates"]["asbaReady"],
        "bankRolesReady": model["infrastructureAggregates"]["configuredBankRoles"],
        "bankRolesTotal": model["infrastructureAggregates"]["requiredBankRoles"],
        "underwritingCoverage": model["underwritingAggregates"]["totalUnderwritingPercentage"],
        "uncoveredShares": model["underwritingAggregates"]["uncoveredShares"],
        "merchantBankerOwnAccountPercentage": model["underwritingAggregates"]["ownAccountPercentage"],
        "marketMakerAppointed": model["marketMakingAggregates"]["marketMakerAppointed"],
        "marketMakingAgreementExecuted": model["marketMakingAggregates"]["agreementExecuted"],
        "marketMakingReservationStatus": (section6.get("marketMakerReservation") or {}).get(
            "allocationStatus", ""
        ),
        "issueOpeningDate": model["programmeAggregates"]["issueOpeningDate"],
        "issueClosingDate": model["programmeAggregates"]["issueClosingDate"],
        "preliminaryTPlus3ListingDate": model["programmeAggregates"]["preliminaryTPlus3ListingDate"],
        "basisStatus": (section7.get("basisOfAllotment") or {}).get("basisPrepared", ""),
        "dematStatus": (section7.get("dematCredit") or {}).get("sharesCredited", ""),
        "listingStatus": (section7.get("listing") or {}).get("listingCompletionStatus", ""),
        "unresolvedPlaceholders": model["finalDocumentAggregates"]["openPlaceholderCount"],
        "inspectionItemsPending": model["finalDocumentAggregates"]["inspectionItemsPending"],
        "issueAgreementsPending": model["finalDocumentAggregates"]["issueAgreementsPending"],
        "advertisementsPending": model["finalDocumentAggregates"]["publicCommunicationsPending"],
        "repositoryReadiness": repository_readiness,
        "reconciliationMismatchCount": model["reconciliation"]["totalMismatchCount"],
        "assessmentConcerns": assessment_concerns,
        "pendingProfessionalConfirmations": assessment["counts"]["pendingProfessionalConfirmation"],
        "assessmentResult": assessment["result"],
        "assessmentResultLabel": assessment["resultLabel"],
        "assessmentSummary": assessment["summary"],
        "recommendedNextActions": recommended_next_actions,
    }
